File: zZ/cpp/convert-sorted-array-to-binary-search-tree.cpp
// 108. Convert Sorted Array to Binary Search Tree
// Difficulty: Medium

// Given an array where elements are sorted in ascending order, convert it to a height balanced BST.

// Time:  O(n)
// Space: O(h)

class Solution {
public:
	TreeNode* sortedArrayToBST (vector<int>& num) {
		return sortedArrayToBST(num.begin(), num.end());
	}

private:
	template<typename It>
	TreeNode* sortedArrayToBST(It first, It last) {
		if (first >= last) {
			return nullptr;
		}
		auto mid = first + (last - first) / 2;
		TreeNode* root = new TreeNode (*mid);
		root->left = sortedArrayToBST(first, mid);
		root->right = sortedArrayToBST(mid + 1, last);
		return root;
	}
};

File: zZ/java/convert-sorted-array-to-binary-search-tree.java
/**
 * Convert Sorted Array to Binary Search Tree
 * 
 * Given an array where elements are sorted in ascending order, convert it to a height balanced BST.
 */

public class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return sortedArrayToBSTHelper(nums, 0, nums.length - 1);
    }

    private sortedArrayToBSTHelper(int[] nums, int left, int right) {
        if (left > right) {
            return null;
        }

        int middle = left + (right - left) / 2;
        TreeNode root = new TreeNode(nums[middle]);
        root.left = sortedArrayToBSTHelper(nums, left, middle - 1);
        root.right = sortedArrayToBSTHelper(nums, middle + 1, right);

        return root;
    }
}

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}
